electron/js/menu/TrayHandler.js

```diff
@@ -0,0 +1,86 @@
+const {app, Menu, nativeImage, Tray} = require('electron');
+const config = require('./../config');
+const lifecycle = require('./../lifecycle');
+const locale = require('./../../locale/locale');
+const path = require('path');
+const windowManager = require('./../window-manager');
+
+function buildTrayMenu() {
+  const contextMenu = Menu.buildFromTemplate([
+    {
+      click: () => windowManager.showPrimaryWindow(),
+      label: locale.getText('trayOpen'),
+    },
+    {
+      click: async () => await lifecycle.quit(),
+      label: locale.getText('trayQuit'),
+    },
+  ]);
+
+  this.trayIcon.on('click', () => windowManager.showPrimaryWindow());
+  this.trayIcon.setContextMenu(contextMenu);
+  this.trayIcon.setToolTip(config.NAME);
+}
+
+function flashApplicationWindow(win, count) {
+  if (win.isFocused() || !count) {
+    win.flashFrame(false);
+  } else if (count > this.lastUnreadCount) {
+    win.flashFrame(true);
+  }
+}
+
+function updateBadgeCount(count) {
+  app.setBadgeCount(count);
+  this.lastUnreadCount = count;
+}
+
+function updateTrayIcon(win, count) {
+  if (this.trayIcon) {
+    const trayImage = count ? this.icons.trayWithBadge : this.icons.tray;
+    this.trayIcon.setImage(trayImage);
+  }
+
+  if (this.hasOverlaySupport) {
+    const overlayImage = count ? this.icons.badge : null;
+    win.setOverlayIcon(overlayImage, locale.getText('unreadMessages'));
+  }
+}
+
+class TrayHandler {
+  constructor() {
+    this.hasOverlaySupport = false;
+    this.lastUnreadCount = 0;
+  }
+
+  initTray(hasOverlaySupport, trayIcon = new Tray(nativeImage.createEmpty())) {
```

Comment:
Why moving this part to another method? This creates *temporal coupling* (the consumer now has to call the methods in the right order else he will have unexpected behaviors)

Reply:
I want the consumer to decide if he/she wants to have a tray icon or not. That's why it got decoupled.

Reply:
Very valid point!